From b03cdd3bd07f19eeac2d65a04361e48ebde31a73 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Thu, 16 Oct 2025 10:08:32 +0200
Subject: [PATCH 435/553] =?UTF-8?q?QCborValue:=20rewrite=20nextUtf32Charac?=
 =?UTF-8?q?ter=20to=20avoid=20narrowing=20char32=5Ft=20=E2=86=92=20char16?=
 =?UTF-8?q?=5Ft?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Ought to fix Clang 21 -Wcharacter-conversion in this function.

Instead of storing the initial *ptr in r.c (a char32_t) and then
narrowing it back to char16_t for the call to
QChar::surrogateToUcs4(), store it in a separate variable and
construct an R only in the return statement.

Also move the len = 1 return from the end of the function to the
(inverted) if (c < 0x80) near the start of the function.

This is less assmebler-esque, so might execute slower (though it's all
values, so the optimizer should have no problem rewirting this to the
old code, if that was faster), but is much easier to read, which is
not the sneezed at in a security-critical component.

Amends d4c7da9a07dc1434692fe08a61ba22c794574c4f.

Pick-to: 6.8 6.5
Change-Id: I1ce610fa80e3d874106c9c5b3a1d50bf99f39732
Reviewed-by: Thiago Macieira <thiago.macieira@intel.com>
Reviewed-by: Ahmad Samir <a.samirh78@gmail.com>
(cherry picked from commit a4d5432fcc182535ce2014695516b482a4a76777)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/serialization/qcborvalue.cpp | 24 ++++++++++++------------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/src/corelib/serialization/qcborvalue.cpp b/src/corelib/serialization/qcborvalue.cpp
index f905b97a164..13d74e591d5 100644
--- a/src/corelib/serialization/qcborvalue.cpp
+++ b/src/corelib/serialization/qcborvalue.cpp
@@ -1102,20 +1102,20 @@ static auto nextUtf32Character(const char16_t *&ptr, const char16_t *end) noexce
     Q_ASSERT(ptr != end);
     struct R {
         char32_t c;
-        qsizetype len = 1;  // in UTF-8 code units (bytes)
-    } r = { *ptr++ };
-
-    if (r.c < 0x0800) {
-        if (r.c >= 0x0080)
-            ++r.len;
-    } else if (!QChar::isHighSurrogate(r.c) || ptr == end) {
-        r.len += 2;
+        qsizetype len;  // in UTF-8 code units (bytes)
+    };
+
+    const char16_t c = *ptr++;
+
+    if (c < 0x0800) {
+        if (c < 0x0080)
+            return R{c, 1};
+        return R{c, 2};
+    } else if (!QChar::isHighSurrogate(c) || ptr == end) {
+        return R{c, 3};
     } else {
-        r.len += 3;
-        r.c = QChar::surrogateToUcs4(r.c, *ptr++);
+        return R{QChar::surrogateToUcs4(c, *ptr++), 4};
     }
-
-    return r;
 }
 
 static qsizetype stringLengthInUtf8(const char16_t *ptr, const char16_t *end) noexcept
-- 
2.51.2

